File: crud/javascript/peliculas.js
// Definición de la clase Peliculas para gestionar nuestro catálogo en memoria
class Peliculas {
  // El constructor inicializa el estado de cada nueva instancia de la clase
  constructor() {
    // Array interno donde se guardarán los objetos de cada película
    this._peliculas = [];
    // Contador para asignar identificadores únicos y secuenciales
    this._nextId = 1;
  }

  // Método para registrar una nueva película en el sistema
  crear(titulo, genero, anio, calificacion) {
    // Validación: si la calificación no está entre 1 y 10, detenemos la ejecución
    if (calificacion < 1.0 || calificacion > 10.0) {
      // Lanzamos un error que podrá ser capturado por quien llame al método
      throw new Error('La calificación debe estar entre 1.0 y 10.0');
    }
    // Creamos el objeto película agrupando las propiedades pasadas como argumentos
    const pelicula = {
      id: this._nextId++, // Asignamos el ID actual y luego lo incrementamos para la próxima vez
      titulo,             // Sintaxis abreviada (shorthand): equivale a titulo: titulo
      genero,
      anio,
      calificacion,
    };
    // Guardamos la nueva película al final del array interno usando push()
    this._peliculas.push(pelicula);
    // Retornamos el objeto creado para confirmar el éxito de la operación
    return pelicula;
  }

  // Método para recuperar el listado completo de películas
  obtenerTodas() {
    // Retornamos una copia del array usando el operador spread (...)
    // Esto protege nuestro array original de modificaciones externas accidentales
    return [...this._peliculas];
  }

  // Método para buscar una película específica mediante su ID
  obtenerPorId(id) {
    // find() busca el primer elemento que cumpla la condición (p.id === id)
    // Si no lo encuentra, find() devuelve undefined, en cuyo caso devolvemos null
    return this._peliculas.find(p => p.id === id) || null;
  }

  // Método para filtrar películas que pertenezcan a un género dado
  obtenerPorGenero(genero) {
    // filter() devuelve un nuevo array con todos los elementos que coincidan con el género
    return this._peliculas.filter(p => p.genero === genero);
  }

  // Método para actualizar los datos de una película existente
  actualizar(id, datos) {
    // Buscamos el índice (posición) de la película en nuestro array
    const index = this._peliculas.findIndex(p => p.id === id);
    // Si el índice es -1, significa que no existe ninguna película con ese ID
    if (index === -1) return null;
    // Combinamos la película actual con los nuevos datos usando el operador spread
    // Las propiedades en 'datos' sobrescribirán a las de la película original
    this._peliculas[index] = { ...this._peliculas[index], ...datos };
    // Retornamos el objeto actualizado
    return this._peliculas[index];
  }

  // Método para eliminar permanentemente una película del sistema
  eliminar(id) {
    // Localizamos de nuevo la posición de la película en el array
    const index = this._peliculas.findIndex(p => p.id === id);
    // Si no se encuentra, retornamos false indicando que no se pudo borrar nada
    if (index === -1) return false;
    // splice() elimina 1 elemento en la posición especificada por 'index'
    this._peliculas.splice(index, 1);
    // Retornamos true para confirmar que la eliminación fue exitosa
    return true;
  }
}

// Exportamos la clase para que pueda ser utilizada en otros archivos (Node.js CommonJS)
module.exports = { Peliculas };

File: crud/python/peliculas.py
# Definición de la clase Peliculas que actuará como nuestro gestor de datos (Simulando una base de datos en memoria)
class Peliculas:
    # Método constructor que se ejecuta al crear una instancia de la clase
    def __init__(self):
        # Inicializamos una lista vacía para almacenar los diccionarios de cada película
        self._peliculas = []
        # Inicializamos un contador para asignar IDs únicos de forma incremental, empezando en 1
        self._next_id = 1

    # Método para crear y añadir una nueva película a nuestro sistema
    def crear(self, titulo, genero, anio, calificacion):
        # Validación de negocio: comprobamos que la calificación esté en el rango permitido (1.0 a 10.0)
        if calificacion < 1.0 or calificacion > 10.0:
            # Si no es válida, lanzamos una excepción de tipo ValueError con un mensaje explicativo
            raise ValueError('La calificación debe estar entre 1.0 y 10.0')
        
        # Creamos un diccionario que representa el objeto película con sus propiedades
        pelicula = {
            'id': self._next_id,          # Asignamos el ID actual almacenado en el contador
            'titulo': titulo,             # Guardamos el título pasado como argumento
            'genero': genero,             # Guardamos el género pasado como argumento
            'anio': anio,                 # Guardamos el año pasado como argumento
            'calificacion': calificacion, # Guardamos la calificación validada
        }
        
        # Incrementamos el contador de IDs para que la siguiente película tenga un ID diferente y único
        self._next_id += 1
        
        # Añadimos el diccionario de la película recién creada a nuestra lista principal
        self._peliculas.append(pelicula)
        
        # Devolvemos el objeto película creado para que quien llamó al método pueda usarlo
        return pelicula

    # Método para obtener el listado completo de películas almacenadas
    def obtener_todas(self):
        # Devolvemos una copia de la lista para evitar que se modifique la lista original desde fuera
        return list(self._peliculas)

    # Método para buscar una película específica utilizando su identificador único (ID)
    def obtener_por_id(self, id):
        # Recorremos cada película dentro de nuestra lista interna
        for p in self._peliculas:
            # Verificamos si el ID de la película actual coincide con el ID buscado
            if p['id'] == id:
                # Si lo encontramos, devolvemos el diccionario de la película inmediatamente
                return p
        # Si terminamos de recorrer la lista y no encontramos ninguna coincidencia, devolvemos None (nulo)
        return None

    # Método para filtrar y obtener películas que pertenezcan a un género específico
    def obtener_por_genero(self, genero):
        # Usamos una "list comprehension" para crear una nueva lista con solo las películas que coincidan
        # Recorre cada 'p' en 'self._peliculas' y la incluye si 'p["genero"]' es igual al género buscado
        return [p for p in self._peliculas if p['genero'] == genero]

    # Método para actualizar la información de una película existente identificada por su ID
    def actualizar(self, id, datos):
        # Recorremos la lista usando enumerate para obtener tanto el índice (i) como la película (p)
        for i, p in enumerate(self._peliculas):
            # Comprobamos si el ID de la película en esta posición es el que queremos actualizar
            if p['id'] == id:
                # Actualizamos la película combinando el diccionario original (**p) con los nuevos datos (**datos)
                # El operador ** desempaqueta los diccionarios; si hay claves repetidas, prevalecen las de 'datos'
                self._peliculas[i] = {**p, **datos}
                # Devolvemos la película ya actualizada
                return self._peliculas[i]
        # Si no encontramos el ID proporcionado, devolvemos None indicando que no se pudo actualizar
        return None

    # Método para eliminar una película del sistema basándose en su ID
    def eliminar(self, id):
        # Recorremos la lista con índice y objeto para poder identificar la posición exacta
        for i, p in enumerate(self._peliculas):
            # Verificamos si encontramos la película con el ID especificado
            if p['id'] == id:
                # Usamos el método pop() de las listas para eliminar el elemento en la posición 'i'
                self._peliculas.pop(i)
                # Retornamos True indicando que la operación de eliminación fue exitosa
                return True
        # Si el bucle termina sin encontrar el ID, retornamos False indicando que no se borró nada
        return False
